File: 3310-remove-methods-from-project/3310-remove-methods-from-project.cpp
class Solution {
public:
    vector<int>ans;
    vector<bool>visited;
    void dfs(int node,vector<vector<int>>& nums){
        visited[node] = true;
        for(auto it:nums[node]){
           if(!visited[it]) dfs(it,nums);
        }
    }
    vector<int> remainingMethods(int n, int k, vector<vector<int>>& invocations) {
        vector<vector<int>> nums(n);
        visited.resize(n,false);
        for(auto it : invocations){
            nums[it[0]].push_back(it[1]);
        }
        dfs(k,nums);
        
        for(auto it : invocations){
            if(!visited[it[0]] && visited[it[1]]){
                for(int i=0;i<n;i++)ans.push_back(i);
                return ans;
            }
        }

        for(int i=0;i<n;i++){
            if(!visited[i])ans.push_back(i);
        }
        return ans;

    }
};

File: 3310-remove-methods-from-project/3310-remove-methods-from-project.kt
class Solution {
    fun remainingMethods( n: Int, k: Int, invocations: Array<IntArray> ): List<Int> {

        val graph = Array(n) { mutableListOf<Int>() }

        for (edge in invocations) {
            graph[edge[0]].add(edge[1])
        }

        val suspicious = BooleanArray(n)

        fun dfs(node: Int) {
            suspicious[node] = true
            for (next in graph[node]) {
                if (!suspicious[next]) {
                    dfs(next)
                }
            }
        }

        dfs(k)

        // If an outside method invokes a suspicious method,
        // nothing can be removed.
        for (edge in invocations) {
            val u = edge[0]
            val v = edge[1]

            if (!suspicious[u] && suspicious[v]) {
                val ans = mutableListOf<Int>()
                for (i in 0 until n) ans.add(i)
                return ans
            }
        }

        val ans = mutableListOf<Int>()

        for (i in 0 until n) {
            if (!suspicious[i]) {
                ans.add(i)
            }
        }

        return ans
    }
}